zwad/aad.py

```diff
@@ -1,49 +1,26 @@
 #!/usr/bin/env python
 
+import argparse
 import click
-import logging
-import os
+import numpy as np
+import pandas as pd
 import sys
 import webbrowser
 
-import numpy as np
-import pandas as pd
+from coniferest.aadforest import AADForest, AADForestAnomalyDetector
+from coniferest.datasets import Label
+from coniferest.pineforest import PineForest, PineForestAnomalyDetector
 
-from ad_examples.common.utils import configure_logger
-from ad_examples.common.gen_samples import get_synthetic_samples
-from ad_examples.aad import aad_globals
-from ad_examples.aad.aad_globals import (
-    AAD_IFOREST, IFOR_SCORE_TYPE_NEG_PATH_LEN, HST_LOG_SCORE_TYPE, AAD_HSTREES, RSF_SCORE_TYPE,
-    AAD_RSFOREST, INIT_UNIF, AAD_CONSTRAINT_TAU_INSTANCE, QUERY_DETERMINISIC, ENSEMBLE_SCORE_LINEAR,
-    AadOpts
-)
-from ad_examples.aad.aad_support import get_aad_model
-from ad_examples.aad.forest_aad_detector import is_forest_detector
-from ad_examples.aad.forest_description import CompactDescriber, MinimumVolumeCoverDescriber, BayesianRulesetsDescriber, get_region_memberships
-from ad_examples.aad.query_model import Query
-
-from zwad.ad.transformation import transform_features as transform_lc_features
 from zwad.utils import load_data
 
-import onnx
-from aad2onnx.convert import to_onnx
-
-"""
-A simple no-frills demo of how to use AAD in an interactive loop.
 
-To execute:
-pythonw -m aad.demo_aad
-"""
-
-logger = logging.getLogger(__name__)
-
-
-class ResultsSink(object):
-    def __init__(self, anomalies_filename, answers_filename):
+class ResultsSink:
+    def __init__(self, anomalies_filename, answers_filename = None):
         self._anomalies = open(anomalies_filename, mode="w", encoding="utf-8")
 
-        self._answers = open(answers_filename, mode="w", encoding="utf-8")
-        self._answers.write("oid,is_anomaly\n")
+        self._answers = open(answers_filename, mode="w", encoding="utf-8") if answers_filename is not None else None
+        if self._answers is not None:
+            self._answers.write("oid,is_anomaly\n")
```

Comment:
I think we need `try ... finally ...` block here, because the second `open` could fail keeping the `self._anomalies` opened

Reply:
Sure.